### PR TITLE
Feature/add character

### PR DESCRIPTION
**Pull request by RomanDevelop**

### CODE REVIEW

File: lib/common/network/api/characters_api.dart [hunk 1]
```diff
@@ -10,18 +10,48 @@ part 'characters_api.g.dart';
...
+
+  @POST('')
+  Future<Characters> addCharacter(
+    @Body() Character character,
+  );
 }
 
 class FakeCharactersApi implements CharactersApi {
+  // variable in FakeCharactersApi for ex fakeCharacters and work with it
+  List<Character> fakeCharacters = [];
```
Comment (arrrtem22) [inline lib/common/network/api/characters_api.dart]:
```suggestion
  List<Character> fakeCharacters = [
     Character(
          title: 'Fake Character 1',
          imageUrl: 'https://via.placeholder.com/600/92c952',
          description: 'This is a fake character description.',
        ),
        Character(
          title: 'Fake Character 2',
          imageUrl: 'https://via.placeholder.com/600/7286a7',
          description: 'Another fake character description.',
        ),
      ];
  ];
```

File: lib/common/network/api/characters_api.dart [hunk 2]
```diff
@@ -10,18 +10,48 @@ part 'characters_api.g.dart';
...
+
+  @override
+  Future<Characters> getCharacters() async {
+    if (fakeCharacters.isEmpty) {
```
Comment (arrrtem22) [inline lib/common/network/api/characters_api.dart]: remove it and leave only delay and variable return
regarding prints, I'll show you how to create logs later
```suggestion
     await Future.delayed(Duration(seconds: 2));
     return fakeCharacters;
```

File: lib/common/network/api/characters_api.dart [hunk 3]
```diff
@@ -10,18 +10,48 @@ part 'characters_api.g.dart';
...
+    print('getCharacters method called');
+    return Characters(characters: fakeCharacters);
+  }
+
   @override
-  Future<Characters> getCharacters() {
-    // TODO: implement ping
-    throw UnimplementedError();
+  Future<Characters> addCharacter(@Body() Character character) async {
+    fakeCharacters.add(character); // adding the new character
```
Comment (arrrtem22) [inline lib/common/network/api/characters_api.dart]: let's add delay for this method and remove return, we should return nothing here

File: lib/feature/add_character/add_character_page.dart [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  const AddCharacterPage({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text('Add Character'),
+      ),
+      body: BlocProvider(
+        create: (context) => CharacterCubit(
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]: check my files how I initialize cubit in BlocProvider, as I remember, I don't create new instances using constructors

File: lib/common/network/api/characters_api.dart (hunk 1 above)
Comment (RomanDevelop) [reply lib/common/network/api/characters_api.dart]: Done!)

File: lib/common/network/api/characters_api.dart (hunk 2 above)
Comment (RomanDevelop) [reply lib/common/network/api/characters_api.dart]: Done!)

File: lib/feature/add_character/add_character_page.dart (hunk 4 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done) recreated)
Mason is the best tool I have ever seen!
Thanks!

File: lib/common/network/api/characters_api.dart (hunk 3 above)
Comment (RomanDevelop) [reply lib/common/network/api/characters_api.dart]: Done

File: lib/feature/home/home_page.dart [hunk 5]
```diff
@@ -69,6 +70,12 @@ class _LargeHomeView extends StatelessWidget {
                 child: state.selected != null
                     ? DetailView(character: state.selected!)
                     : const Center(child: Text('Select character'))),
+            FloatingActionButton(
+              onPressed: () {
+                GoRouter.of(context).go('/addCharacter');
```
Comment (arrrtem22) [inline lib/feature/home/home_page.dart]: navigation should be the same like here https://github.com/arrrtem22/character_viewer/pull/12/files#diff-7d0f01b792ad3bfd3412eb49c3903abf3231bede771b500d6d32f84b37c89c12R106

go_router also uses code generation so do not forget to run make gen after creating new route class

File: lib/common/service/character_service.dart [hunk 6]
```diff
@@ -10,4 +10,12 @@ class CharactersService {
   final CharactersApi _charactersApi;
 
   Future<Characters> getCharacters() => _charactersApi.getCharacters();
+
+  Future<void> addCharacter(Character newCharacter) async {
+    try {
+      await _charactersApi.addCharacter(newCharacter);
+    } catch (e) {
+      throw Exception('Error during adding the character: $e');
```
Comment (arrrtem22) [inline lib/common/service/character_service.dart]: add todo here, maybe we'll add it later
but now I don't think that it's necessary
```suggestion
      // TODO emit failure state
      throw Exception('Error during adding the character: $e');
```

File: lib/feature/add_character/add_character_page.dart [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    } else if (state is AddCharacterStateFailure) {
+      return _buildErrorUI(context, state.error);
+    } else {
+      return const Center(
+        child: CircularProgressIndicator(),
+      );
+    }
+  }
+
+  Widget _buildInitialUI(BuildContext context) {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]: change helper methods to widgets

https://www.youtube.com/watch?v=IOyq-eTRhvo&ab_channel=Flutter

File: lib/feature/add_character/add_character_page.dart [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+          title: const Text('Add Character'),
+        ),
+        body: _AddCharacterView(),
+      ),
+    );
+  }
+}
+
+class _AddCharacterView extends StatelessWidget {
+  final TextEditingController titleController = TextEditingController();
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart:32]: we shouldn't use controllers in StatelessWidget, we either do not use them or they should be in StatefulWidget

File: lib/feature/add_character/add_character_page.dart [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            'Character Description',
+          ),
+          const SizedBox(height: 20),
+          ElevatedButton(
+            onPressed: () {
+              final title = titleController.text.trim();
+              final imageUrl = imageUrlController.text.trim();
+              final description = descriptionController.text.trim();
+
+              if (title.isNotEmpty &&
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]: that's business logic that should be in bloc and bloc emits states to show error or success or something different

File: lib/feature/add_character/add_character_page.dart [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        ],
+      ),
+    );
+  }
+
+  Widget _buildNeumorphicTextField(
+      TextEditingController controller, String labelText) {
+    return Container(
+      decoration: BoxDecoration(
+        color: Colors.grey[200], // Цвет фона текстового поля
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]: wow, I don't think that we should use Russian to write some comments in code. Please remove it or use English

File: lib/feature/add_character/add_character_page.dart [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+          Text(
+            'Error: $error',
+            style: const TextStyle(fontSize: 18, color: Colors.red),
+          ),
+          const SizedBox(height: 10),
+          ElevatedButton(
+            onPressed: () {
+              context
+                  .read<AddCharacterCubit>()
+                  .addCharacter(title: '', imageUrl: '', description: '');
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]: is it okay that we don't have parameters here?

File: lib/feature/add_character/cubit/add_character_cubit.dart [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+@injectable
+class AddCharacterCubit extends Cubit<AddCharacterState> {
+  final CharactersService _charactersService;
+
+  AddCharacterCubit(this._charactersService)
+      : super(const AddCharacterState.initial());
+
+  void addCharacter({
+    required String title,
```
Comment (arrrtem22) [inline lib/feature/add_character/cubit/add_character_cubit.dart]: it's not critical but if you wanna do it better you can create Character in UI when you call addCharacter method and pass only one parameter (Character character)

File: lib/feature/add_character/cubit/add_character_cubit.dart [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    required String description,
+  }) {
+    try {
+      final newCharacter = Character(
+        title: title,
+        imageUrl: imageUrl,
+        description: description,
+      );
+
+      _charactersService.addCharacter(newCharacter);
```
Comment (arrrtem22) [inline lib/feature/add_character/cubit/add_character_cubit.dart]:
```suggestion
      await _charactersService.addCharacter(newCharacter);
```

File: lib/common/service/character_service.dart (hunk 6 above)
Comment (arrrtem22) [reply lib/common/service/character_service.dart]: oh no, I thought it's bloc class then we don't need try catch at all

File: package/cached_repository/lib/resource.dart [hunk 14]
```diff
@@ -49,18 +49,23 @@ class Resource<T> {
           state == other.state &&
           _compareData(data, other.data) &&
           message == other.message &&
-          error == other.error;
+          error == other.error &&
+          stackTrace == other.stackTrace;
 
   bool _compareData(dynamic v1, dynamic v2) =>
-      v1 is List && v2 is List ? listEquals(v1, v2) : v1 == v2;
+      v1 is List && v2 is List ? const ListEquality().equals(v1, v2) : v1 == v2;
```
Comment (arrrtem22) [inline package/cached_repository/lib/resource.dart:56]: what is the difference?

File: lib/feature/home/home_page.dart (hunk 5 above)
Comment (RomanDevelop) [reply lib/feature/home/home_page.dart]: Ok
Done

File: lib/common/service/character_service.dart (hunk 6 above)
Comment (RomanDevelop) [reply lib/common/service/character_service.dart]: Ok
Clear. I have deleted it.
As I have understood this one is as interface? Is it correct?

File: lib/feature/add_character/add_character_page.dart (hunk 7 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 8 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart:32]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 9 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Removed

File: lib/feature/add_character/add_character_page.dart (hunk 10 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 11 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done
Added ones)

File: lib/feature/add_character/cubit/add_character_cubit.dart (hunk 12 above)
Comment (RomanDevelop) [reply lib/feature/add_character/cubit/add_character_cubit.dart]: Done
Removed

File: package/cached_repository/lib/resource.dart (hunk 14 above)
Comment (RomanDevelop) [reply package/cached_repository/lib/resource.dart:56]: Both snippets are equivalent and archive the same result. In the project the first one was with false)) for package))

File: lib/feature/add_character/cubit/add_character_cubit.dart (hunk 13 above)
Comment (RomanDevelop) [reply lib/feature/add_character/cubit/add_character_cubit.dart]: Need advice!))
If I have changed to 'async await' as a result the  snake bar doesn't work (

File: lib/feature/add_character/add_character_page.dart [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+                  );
+            },
+          ),
+        );
+      },
+    );
+  }
+}
+
+class BuildUI extends StatefulWidget {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart:80]: Change to stless

File: lib/feature/add_character/add_character_page.dart [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+          ),
+          const SizedBox(height: 20),
+          AddCharacterButtonWidget(onPressed: widget.onPressed),
+        ],
+      ),
+    );
+  }
+}
+
+class NeumorphicTextFieldWidget extends StatefulWidget {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]:
```suggestion
class _NeumorphicTextFieldWidget extends StatelessWidget {
```

File: lib/feature/add_character/add_character_page.dart (hunk 15 above)
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]:
```suggestion
class _BuildUI extends StatelessWidget {
```

File: lib/feature/add_character/add_character_page.dart [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+          labelText: widget.labelText,
+          contentPadding: const EdgeInsets.symmetric(horizontal: 16),
+          border: InputBorder.none,
+        ),
+      ),
+    );
+  }
+}
+
+class AddCharacterButtonWidget extends StatelessWidget {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]:
```suggestion
class _AddCharacterButtonWidget extends StatelessWidget {
```

File: lib/feature/add_character/add_character_page.dart (hunk 16 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 15 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 17 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 16 above)
Comment (arrrtem22) [reply lib/feature/add_character/add_character_page.dart]: I don't see these changes

File: lib/feature/add_character/add_character_page.dart (hunk 15 above)
Comment (arrrtem22) [reply lib/feature/add_character/add_character_page.dart]: Did you push it?

File: lib/feature/add_character/add_character_page.dart [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+          labelText: labelText,
+          contentPadding: const EdgeInsets.symmetric(horizontal: 16),
+          border: InputBorder.none,
+        ),
+      ),
+    );
+  }
+}
+
+class AddCharacterButtonWidget extends StatefulWidget {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]: still stateful

File: lib/common/network/api/characters_api.dart [hunk 19]
```diff
@@ -46,9 +46,9 @@ class FakeCharactersApi implements CharactersApi {
   }
 
   @override
-  Future<Characters> addCharacter(@Body() Character newCharacter) async {
+  Future<void> addCharacter(@Body() Character newCharacter) async {
     await Future.delayed(const Duration(seconds: 2));
     fakeCharacters.add(newCharacter);
-    return Characters(characters: fakeCharacters);
+    print('all $fakeCharacters');
```
Comment (arrrtem22) [inline lib/common/network/api/characters_api.dart]: remove print, we add prints only for debugging

File: lib/common/network/api/characters_api.dart [hunk 20]
```diff
@@ -10,18 +12,43 @@ part 'characters_api.g.dart';
...
   @GET('')
   Future<Characters> getCharacters();
+
+  @POST('')
+  Future<void> addCharacter(
+    @Body() Character newCharacter,
+  );
 }
 
 class FakeCharactersApi implements CharactersApi {
```
Comment (arrrtem22) [inline lib/common/network/api/characters_api.dart]:
```suggestion
class _FakeCharactersApi implements CharactersApi {
```

File: lib/feature/add_character/add_character_page.dart [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+                  );
+            },
+          ),
+        );
+      },
+    );
+  }
+}
+
+class BuildUI extends StatelessWidget {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]:
```suggestion
class _BuildUI extends StatelessWidget {
```

File: lib/feature/add_character/add_character_page.dart [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+          ),
+          const SizedBox(height: 20),
+          AddCharacterButtonWidget(onPressed: onPressed),
+        ],
+      ),
+    );
+  }
+}
+
+class NeumorphicTextFieldWidget extends StatelessWidget {
```
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]:
```suggestion
class _NeumorphicTextFieldWidget extends StatelessWidget {
```

File: lib/feature/add_character/add_character_page.dart (hunk 18 above)
Comment (arrrtem22) [inline lib/feature/add_character/add_character_page.dart]:
```suggestion
class _AddCharacterButtonWidget extends StatefulWidget {
```

File: lib/feature/home/home_page.dart (hunk 5 above)
Comment (arrrtem22) [inline lib/feature/home/home_page.dart]:
```suggestion
                AddCharacterRoute().go(context);
```

File: lib/feature/add_character/add_character_page.dart (hunk 18 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/common/network/api/characters_api.dart (hunk 19 above)
Comment (RomanDevelop) [reply lib/common/network/api/characters_api.dart]: Done

File: lib/common/network/api/characters_api.dart (hunk 20 above)
Comment (RomanDevelop) [reply lib/common/network/api/characters_api.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 21 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 22 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/add_character/add_character_page.dart (hunk 18 above)
Comment (RomanDevelop) [reply lib/feature/add_character/add_character_page.dart]: Done

File: lib/feature/home/home_page.dart (hunk 5 above)
Comment (RomanDevelop) [reply lib/feature/home/home_page.dart]: Done

**Review by arrrtem22 — APPROVED** [review]

**Pull request merged**